### PR TITLE
JDK-8293346: Add NoThreadCurrentMark to simulate non-attached threads

### PR DESCRIPTION
**Pull request by tstuefe**
When working on [JDK-8293344](https://bugs.openjdk.org/browse/JDK-8293344), I needed a way to verify that I had removed all offending usages of ResourceArea memory inside an extend.

For that, a NoThreadCurrentMark was useful. It temporarily sets the current Thread* to NULL for the extend. Any use of Thread::current below that mark will now crash or assert.

This can be used for two things:
- guard code that is supposed to be safe for non-attached threads (os::malloc, stack printing, UL etc) against accidental usage of Thread::current() (e.g. resource area allocation)
- in gtests, simulate a non-attached thread to cover code that behaves differently with Thread::current()==NULL.

This patch:

- Introduces the debug-only `NoThreadCurrentMark`
- Adds a gtest for it
- Adds it to guard dwarf-parsing-based stack printing, which had not been safe due to use of ResourceArea, fixed with  [JDK-8293344](https://bugs.openjdk.org/browse/JDK-8293344)
- Adds it to `os::malloc()`, `os::realloc()` and `os::free()`, since those can certainly be called without a current thread.
- Replaces the test-local mark in the SafeFetch gtests that did a similar thing
- I may add and fix up other users later

I also had to change Library-based TLS initialization such that it runs at C++ dynamic initialization time. That is because the `NoThreadCurrentMark` needs to set both Library-based TLS and C++ TLS slots, and having library-based TLS not available before VM initialization made the code too complex. This also means we can remove the explicit TLS initialization from create_vm.

Finally, while I was here, I also fixed the error printing in library based TLS (since the pthread_key_xxx functions don't modify errno).

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293346](https://bugs.openjdk.org/browse/JDK-8293346): Add NoThreadCurrentMark to simulate non-attached threads

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10178/head:pull/10178` \
`$ git checkout pull/10178`

Update a local copy of the PR: \
`$ git checkout pull/10178` \
`$ git pull https://git.openjdk.org/jdk pull/10178/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10178`

View PR using the GUI difftool: \
`$ git pr show -t 10178`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10178.diff">https://git.openjdk.org/jdk/pull/10178.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Sorry Thomas but this seems far too disruptive and complex just to support the usecase you outlined.

**Comment by dholmes-ora** [comment]
As you are simulating non-attached thread by attached threads why can't use just have a debug-only field in Thread that is set/cleared by the NTCM, and have Thread::current examine that field?

**Comment by tstuefe** [comment]
> Sorry Thomas but this seems far too disruptive and complex just to support the usecase you outlined.

I was afraid you would dislike the changes to TLS initialization.

> As you are simulating non-attached thread by attached threads why can't use just have a debug-only field in Thread that is set/cleared by the NTCM, and have Thread::current examine that field?

I can try this. This would be simpler. Has to support nesting though. Lets see.

**Comment by dholmes-ora** [comment]
A counter rather than a bit allows nesting easily enough. As this is just a debugging/testing aid it really shouldn't disrupt product code IMO. Thanks.

**Comment by tstuefe** [comment]
New solution, based on a "hide thread current" flag inside Thread.

I'm not a big fan tbh. I liked my first approach, switching thread current to NULL, better. Modulo the TLS initialization changes, those I admit were too complex.

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -647,21 +655,29 @@ inline Thread* Thread::current() {
...
 
 inline Thread* Thread::current_or_null_safe() {
+  Thread* t = nullptr;
   if (ThreadLocalStorage::is_initialized()) {
-    return ThreadLocalStorage::thread();
+    t = ThreadLocalStorage::thread();
+  }
+  if (t != nullptr && t->_thread_current_disabled) {
+    t = nullptr;
   }
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:679]: This needs to be in `ifdef ASSERT`

File: src/hotspot/share/runtime/thread.hpp [hunk 2]
```diff
@@ -647,21 +655,29 @@ inline Thread* Thread::current() {
...
 #else
   if (ThreadLocalStorage::is_initialized()) {
-    return ThreadLocalStorage::thread();
+    t = ThreadLocalStorage::thread();
   }
-  return NULL;
 #endif
+  if (t != nullptr && t->_thread_current_disabled) {
+    t = nullptr;
+  }
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:668]: This needs to be in `ifdef ASSERT`

File: src/hotspot/share/runtime/thread.inline.hpp [hunk 3]
```diff
@@ -91,4 +91,27 @@ inline WXMode Thread::enable_wx(WXMode new_state) {
 }
 #endif // __APPLE__ && AARCH64
 
+
+#ifdef ASSERT
+// Mark disables Thread::current by setting it to NULL for the extend. Can be
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.inline.hpp:96]: s/extend/extent/

File: src/hotspot/share/runtime/thread.inline.hpp [hunk 4]
```diff
@@ -91,4 +91,27 @@ inline WXMode Thread::enable_wx(WXMode new_state) {
...
+// used to simulate running code in non-attached threads, or to guard code
+// against accidental usage of features that depend on Thread::current() (e.g.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.inline.hpp:98]: So the "guard" aspect of this is that we expect to crash if the code tries to use the result of `Thread::current()`?

**Review by dholmes-ora — COMMENTED** [review]
This version of the code seems fine in terms of basic functionality.

I'm less convinced about the actual usefulness of this, but as it is debug only code ... My main query here is why some of this code should ever be being executed by non-attached threads? Conditional pre-VM-init usage? (If it unconditional then we don't need to simulate anything as we will have a NULL current thread.)

**Comment by tstuefe** [comment]
I withdraw this since I don't have time to work on it. I think the idea was sound, but ultimately its not that important.

**Pull request closed without merging**